synapse/storage/databases/main/roommember.py

```diff
@@ -730,19 +749,75 @@ async def get_joined_hosts(self, room_id: str, state_entry):
+        state_entry: "_StateCacheEntry",
+    ) -> FrozenSet[str]:
         # We don't use `state_group`, its there so that we can cache based
         # on it. However, its important that its never None, since two current_state's
         # with a state_group of None are likely to be different.
         assert state_group is not None
 
+        # We use a secondary cache of previous work to allow us to build up the
+        # joined hosts for the given state group based on previous state groups.
+        #
+        # We cache one object per room containing the results of the last state
+        # group we got joined hosts for, with the idea being that generally
+        # `get_joined_hosts` with the "current" state group for the room.
         cache = await self._get_joined_hosts_cache(room_id)
-        return await cache.get_destinations(state_entry)
+
+        # If the state group in the cache matches then its a no-op.
+        if state_entry.state_group == cache.state_group:
+            return frozenset(cache.hosts_to_joined_users)
+
+        # Since we'll mutate the cache we need to lock.
+        with (await self._joined_host_linearizer.queue(room_id)):
+            if state_entry.state_group == cache.state_group:
+                # Same state group, so nothing to do
+                pass
+            elif state_entry.prev_group == cache.state_group:
+                # The cache work is for the previous state group, so we work out
+                # the delta.
+                for (typ, state_key), event_id in state_entry.delta_ids.items():
+                    if typ != EventTypes.Member:
+                        continue
+
+                    host = intern_string(get_domain_from_id(state_key))
+                    user_id = state_key
+                    known_joins = cache.hosts_to_joined_users.setdefault(host, set())
+
+                    event = await self.get_event(event_id)
+                    if event.membership == Membership.JOIN:
+                        known_joins.add(user_id)
+                    else:
+                        known_joins.discard(user_id)
+
+                        if not known_joins:
+                            cache.hosts_to_joined_users.pop(host, None)
+            else:
+                # The cache doesn't match the state group or prev state group,
+                # so we calculate the result from first principles.
+                joined_users = await self.get_joined_users_from_state(
+                    room_id, state_entry
+                )
+
+                cache.hosts_to_joined_users = {}
+                for user_id in joined_users:
+                    host = intern_string(get_domain_from_id(user_id))
+                    cache.hosts_to_joined_users.setdefault(host, set()).add(user_id)
+
+            if state_entry.state_group:
+                cache.state_group = state_entry.state_group
+            else:
+                cache.state_group = object()
```

Comment:
`state_entry.state_group` can be `None`, it's `state_group` that can't be. I think it would be equivalent to always do `cache.state_group = state_group`.

OTOH, I'm thinking that we should probably *not* overwrite the cache if the state group is `object()`, as by definition nothing will be able to use it.

Reply:
I assume you mean "if `state_entry.state_group` is None", and I'm minded to agree.

Reply:
I think that will take a little more rejigging of things, so I'm going to do it as a separate PR.

Reply:
sgtm